tests/timelord/test_new_peak.py

```diff
@@ -43,103 +46,344 @@ async def test_timelord_new_peak_basic(
+                # new unknown peak, weight less then curr peak
+                blocks_2 = bt.get_consecutive_blocks(1, blocks)
+                await _validate_and_add_block(b2, blocks_2[-1])
+                await timelord_api.new_peak_timelord(timelord_peak_from_block(b2, blocks_2[-1]))
+                assert timelord_api.timelord.last_state.last_weight == blocks_1[-1].weight
+                assert timelord_api.timelord.last_state.total_iters == blocks_1[-1].reward_chain_block.total_iters
 
     @pytest.mark.anyio
-    async def test_timelord_new_peak_heavier_unfinished(
+    async def test_timelord_new_peak_unfinished_not_orphaned(
         self, bt: BlockTools, timelord: Tuple[TimelordAPI, ChiaServer], default_1000_blocks: List[FullBlock]
     ) -> None:
+        async with create_blockchain(bt.constants, 2) as (b1, db_wrapper1):
+            timelord_api, _ = timelord
+            for block in default_1000_blocks:
+                await _validate_and_add_block(b1, block)
+
+            peak = timelord_peak_from_block(b1, default_1000_blocks[-1])
+            assert peak is not None
+            assert timelord_api.timelord.new_peak is None
+            await timelord_api.new_peak_timelord(peak)
+            assert timelord_api.timelord.new_peak is not None
+            assert timelord_api.timelord.new_peak.reward_chain_block.get_hash() == peak.reward_chain_block.get_hash()
+
+            # make two new blocks on tip
+            blocks_1 = bt.get_consecutive_blocks(2, default_1000_blocks)
+            block_1 = blocks_1[-2]
+            block_2 = blocks_1[-1]
+            await _validate_and_add_block(b1, block_1)
+            await _validate_and_add_block(b1, block_2)
+
+            block_record = b1.block_record(block_2.header_hash)
+            sub_slot_iters, difficulty = get_next_sub_slot_iters_and_difficulty(
+                bt.constants, len(block_1.finished_sub_slots) > 0, b1.block_record(block_1.prev_header_hash), b1
+            )
+
+            timelord_unf_block = timelord_protocol.NewUnfinishedBlockTimelord(
+                block_1.reward_chain_block.get_unfinished(),
+                difficulty,
+                sub_slot_iters,
+                block_1.foliage,
+                next_sub_epoch_summary(bt.constants, b1, block_record.required_iters, block_1, True),
+                await get_rc_prev(b1, block_1),
+            )
+            await timelord_api.new_unfinished_block_timelord(timelord_unf_block)
+
+            assert timelord_api.timelord.unfinished_blocks[-1].get_hash() == timelord_unf_block.get_hash()
+            new_peak = timelord_peak_from_block(b1, block_2)
+            assert timelord_unf_block.reward_chain_block.total_iters <= new_peak.reward_chain_block.total_iters
+            await timelord_api.new_peak_timelord(new_peak)
+
+            await time_out_assert(60, peak_new_peak_is_none, True, timelord_api)
```

Comment:
I would merge this assertion and the next one below, since the behavior is not intuitive at all. Something like this:

```
def last_state_matches_hash(timelord: TimelordAPI, expected_hash: bytes32) -> bool:
    if timelord.timelord.new_peak is not None:
        # `new_peak_timelord` was not processed yet.
        return False
    return timelord.timelord.last_state.peak.reward_chain_block.get_hash() == expected_hash
```

Similarly with the code below. Otherwise, add a comment explaining `last_state` is not calculated correctly until `new_peak` gets to `None`, because `_handle_new_peak` didn't execute yet.